ssh-oqs.c

```diff
@@ -958,22 +1155,267 @@ static const struct sshkey_impl_funcs sshkey_rsa3072_sphincssha2128fsimple_funcs
+  /* .size = */ ssh_generic_size,
+  /* .alloc = */ ssh_generic_alloc,
+  /* .cleanup = */ ssh_generic_cleanup,
+  /* .equal = */ ssh_generic_equal,
+  /* .ssh_serialize_public = */ ssh_generic_serialize_public,
+  /* .ssh_deserialize_public = */ ssh_generic_deserialize_public,
+  /* .ssh_serialize_private = */ ssh_generic_serialize_private,
+  /* .ssh_deserialize_private = */ ssh_generic_deserialize_private,
+  /* .generate = */ ssh_generic_generate,
+  /* .copy_public = */ ssh_generic_copy_public,
+  /* .sign = */ ssh_generic_sign,
+  /* .verify = */ ssh_generic_verify,
+};
+
+const struct sshkey_impl sshkey_ecdsanistp256_sphincssha2128fsimple_impl = {
+  /* .name = */ "ssh-ecdsa-nistp256-sphincssha2128fsimple",
+  /* .shortname = */ "ECDSA_NISTP256_SPHINCSSHA2128FSIMPLE",
+  /* .sigalg = */ NULL,
+  /* .type = */ KEY_ECDSA_NISTP256_SPHINCS_SHA2_128F_SIMPLE,
+  /* .nid = */ NID_X9_62_prime256v1,
+  /* .cert = */ 0,
+  /* .sigonly = */ 0,
+  /* .keybits = */ 256, // TODO - What should be here?
+  /* .funcs = */ &sshkey_ecdsanistp256_sphincssha2128fsimple_funcs,
+};
+static const struct sshkey_impl_funcs sshkey_ecdsanistp521_sphincssha2256fsimple_funcs = {
+  /* .size = */ ssh_generic_size,
+  /* .alloc = */ ssh_generic_alloc,
+  /* .cleanup = */ ssh_generic_cleanup,
+  /* .equal = */ ssh_generic_equal,
+  /* .ssh_serialize_public = */ ssh_generic_serialize_public,
+  /* .ssh_deserialize_public = */ ssh_generic_deserialize_public,
+  /* .ssh_serialize_private = */ ssh_generic_serialize_private,
+  /* .ssh_deserialize_private = */ ssh_generic_deserialize_private,
+  /* .generate = */ ssh_generic_generate,
+  /* .copy_public = */ ssh_generic_copy_public,
+  /* .sign = */ ssh_generic_sign,
+  /* .verify = */ ssh_generic_verify,
+};
+
+const struct sshkey_impl sshkey_ecdsanistp521_sphincssha2256fsimple_impl = {
+  /* .name = */ "ssh-ecdsa-nistp521-sphincssha2256fsimple",
+  /* .shortname = */ "ECDSA_NISTP521_SPHINCSSHA2256FSIMPLE",
+  /* .sigalg = */ NULL,
+  /* .type = */ KEY_ECDSA_NISTP521_SPHINCS_SHA2_256F_SIMPLE,
+  /* .nid = */ NID_secp521r1,
+  /* .cert = */ 0,
+  /* .sigonly = */ 0,
+  /* .keybits = */ 256, // TODO - What should be here?
+  /* .funcs = */ &sshkey_ecdsanistp521_sphincssha2256fsimple_funcs,
+};
+#endif /* OPENSSL_HAS_ECC */
 #endif /* WITH_OPENSSL */
 ///// OQS_TEMPLATE_FRAGMENT_DEFINE_SIG_FUNCTIONS_END
+
+
+const struct sshkey_impl *oqs_classical_sshkey_impl(const struct sshkey *k)
+{
+  const struct sshkey_impl *impl = NULL;
+  switch(k->type) {
+    CASE_KEY_RSA_HYBRID:
+      impl = &sshkey_rsa_impl;
+      break;
+    CASE_KEY_ECDSA_HYBRID:
+      // Behind the P-256 impl struct is a generic ECDSA implementation which
+      // multiplexes off of either the bits or key->nid parameters passed into
+      // the interface. This behavior is in-line with the "normal" ECDSA code.
+      impl = &sshkey_ecdsa_nistp256_impl;
+      break;
+  }
+  return impl;
```

Comment:
Ah, now I understand why you don't do specific "...->func != NULL" checks: These two function sets are static and always guaranteed to be "OK" in this sense. May be sensible to add a comment here, just in case someone in the future adds another "impl" operating differently.